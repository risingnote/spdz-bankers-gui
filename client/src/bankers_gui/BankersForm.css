.bankersForm {
  margin-right: 3rem;
  text-align: left;
}

.bankersForm label {
  display: block;
}

.bankersForm input {
  display: block;
  margin-bottom: 10px;
}

.diners {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2rem;
}

.atable {
  background: lightyellow;
  border: thin solid grey;
  width: 18rem;
  height: 18rem;
  border-radius: 50%;
  position: relative;
}

.aplace {
  background: white;
  border: thin solid grey;
  width: 4rem;
  height: 4rem;
  border-radius: 10%;
  position: absolute;
  /* centre in container */
  top: 50%;
  left: 50%;
  margin: -2rem 0 0 -2rem;
  /* centre contained text */
  text-align: center;
  vertical-align: middle;
  line-height: 4rem;
}

/* Translation is 
 * (radius.sinØ, -radiuscosØ)
*/