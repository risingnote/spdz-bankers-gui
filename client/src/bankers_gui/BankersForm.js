/**
 * Responsible for managing local state on form before submit.
 */
import React, { Component } from 'react'

import './BankersForm.css'

class BankersForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      bonus: '',
      participantName: ''
    }
    this.handleBonusChange = this.handleBonusChange.bind(this)
    this.handleNameChange = this.handleNameChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleBonusChange(event) {
    this.setState({bonus: event.target.value})
  }

  handleNameChange(event) {
    this.setState({participantName: event.target.value})
  }

  handleSubmit(event) {
    this.props.submitBonus(this.state.participantName, this.state.bonus)
    this.setState({bonus: '', participantName: ''})
    event.preventDefault()
  }

  render() {
    const disableSubmit = this.props.proxiesConnected && !this.props.joinedName  ? '' : 'disabled'
    const statusMessage = () => {
      if (!this.props.proxiesConnected) return 'You must be connected to all SPDZ proxies.'
      else if (this.props.joinedName) return `You have joined the meal as ${this.props.joinedName}.`
      else return ''
    }

    return (
        <form className="BankersForm" onSubmit={this.handleSubmit}>
          <h4>Bankers Celebration Dinner</h4>
          <p>...but who should pay?</p>        
          <label htmlFor="joinName">Join meal as</label>
          <input type="text" id="joinName" value={this.state.participantName} onChange={this.handleNameChange} disabled={disableSubmit}/>
          <label htmlFor="bonusValue">Bonus</label>
          <input type="text" id="bonusValue" value={this.state.bonus} onChange={this.handleBonusChange} disabled={disableSubmit}/>
          <input type="submit" value="Send" disabled={disableSubmit}/>
          <p className='smallText'>{statusMessage()}</p>
        </form>
    )
  }
}

BankersForm.propTypes = {
  submitBonus: React.PropTypes.func.isRequired,
  proxiesConnected: React.PropTypes.bool.isRequired,
  joinedName: React.PropTypes.string
}

export default BankersForm
