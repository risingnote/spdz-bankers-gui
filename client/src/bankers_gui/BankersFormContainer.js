/**
 * Responsible for managing behaviour around form submission for the bankers GUI.
 * Manages websocket connections to maintain list of diners.
 * Polls for a result once meal is joined.
 * Interacts with SPDZ to send shares of data and poll for result.
 */
import React, { Component, PropTypes } from 'react'
import Io from 'socket.io-client'

import { connectToProxies, disconnectFromProxies, allProxiesConnected, 
         sendInputsWithShares, retrieveWinnerClientId, NoContentError } from 'spdz-gui-lib'

import BankersForm from './BankersForm'

class BankersFormContainer extends Component {
  constructor (props) {
    super(props)
    this.state = {
      dinersList: [],
      socket: undefined,
      winningClientId: undefined
    }
    this.resultTimerId = undefined
    this.handleSubmitEntry = this.handleSubmitEntry.bind(this)
    this.resetGame = this.resetGame.bind(this)
    this.joinMeal = this.joinMeal.bind(this)
    this.pollForResult = this.pollForResult.bind(this)
  }

  /**
   * At startup get list of diners who have already joined the meal.
   */
  componentDidMount() {
    const socket = Io('/diners', {path: '/bankers/socket.io'})
    this.setState({socket: socket})

    // Expect list of {name: <somename>, publicKey: <key>}
    socket.on('diners', (msg) => {
      this.setState({dinersList: msg})
      if (msg.length === 0) { //indicates game reset as all diners removed
        this.setState({winningClientId: undefined})
        this.props.changeToDiners(msg, undefined)        
        console.log('All game players reset.')       
      } else {
        this.props.changeToDiners(msg, this.state.winningClientId)        
      }
    });
  }

  componentWillUnmount() {
    // this.state.socket.on('disconnect', () => {})
    if (this.resultTimerId !== undefined) {
      clearInterval(this.resultTimerId)
      this.resultTimerId = undefined
    } 
    disconnectFromProxies(this.props.spdzProxyServerList.map( spdzProxy => spdzProxy.get('url')), 
                          this.props.spdzApiRoot, this.props.clientPublicKey)
  }

  /**
   * Start an interval timer to poll the SPDZ proxies for the winner of the computation.
   * If successfully get result:
   *   Disconnect from SPDZ engines and notify change of connection status.
   */
  pollForResult(spdzProxyServerList, spdzApiRoot, clientPublicKey) {
    this.resultTimerId = setInterval(() => {
      retrieveWinnerClientId(spdzProxyServerList, spdzApiRoot, clientPublicKey)
      .then( winningClientId => {
        this.setState({winningClientId: winningClientId})
        this.props.changeToDiners(this.state.dinersList, winningClientId)
        console.log('Got winning client of ', winningClientId)        
        clearInterval(this.resultTimerId)
      })
      .then( () => {
        return disconnectFromProxies(this.props.spdzProxyServerList.map( spdzProxy => spdzProxy.url ), 
                          this.props.spdzApiRoot, this.props.clientPublicKey)
      })
      .then( (values) => {
        this.props.updateConnectionStatus(values)
      })
      .catch( err => {
        if (err instanceof NoContentError) {
          // Keep going, data should turn up.
          console.log('No result data yet.')
        }
        else {
          console.log('Problem when polling for result ', err)
          clearInterval(this.resultTimerId)        
        }
      })
    }, 2000)
  }

  /**
   * Join meal by sending websocket message, wrapped as promise.
   */
  joinMeal(name, clientPublicKey) {
    const that = this
    return new Promise(function(resolve, reject) {
      that.state.socket.emit('joinMeal', {name: name, publicKey: clientPublicKey}, function(error) {
          if (error) {
            reject(error)
          } else {
            resolve()
          }
      })
    })
  } 

  /**
   * Reset game by sending websocket message. No attempt to reset SPDZ parties.
   */
  resetGame() {
    this.state.socket.emit('resetGame', function(error) {
        if (error) {
          console.log('Unable to reset game.', error)
        }
    })
  } 

  /**
   * Join game by:
   *   Running connection to SPDZ engines and notifying change of connection status.
   *   If all OK join meal with a name
   *   If all OK send bonus as shares to all SPDZ proxies.
   *   Start timer polling for result.
   */
  handleSubmitEntry(name, bonus, finished) {
    connectToProxies(this.props.spdzProxyServerList.map( spdzProxy => spdzProxy.url ), 
                                this.props.spdzApiRoot, this.props.clientPublicKey)
      .then( (values) => {
        this.props.updateConnectionStatus(values)
        if (allProxiesConnected(values)) {
          return 
        }
        else {
          return Promise.reject(new Error('Unable to connect to all Spdz Proxy Servers'))
        }
      })
      .then( () => {
        return sendInputsWithShares([bonus, (finished ? 1 : 0)], true, this.props.spdzProxyServerList, 
              this.props.spdzApiRoot, this.props.clientPublicKey, 500)
      })
      .then( () => {
        return this.joinMeal(name, this.props.clientPublicKey)
      })
      .then( () => {
        this.pollForResult(this.props.spdzProxyServerList, this.props.spdzApiRoot, 
              this.props.clientPublicKey)
      })
      .catch((ex) => {
        console.log("Unable to join meal.", ex)
        this.setState({connectionProblem: 'Unable to successfully send bonus to SPDZ parties.'})
      })
  }

  render() {
    const myEntry = this.state.dinersList.find(diner => diner.publicKey === this.props.clientPublicKey)
    const joinedName = (myEntry !== undefined ? myEntry.name : undefined)
    const winnerChosen = this.state.winningClientId !== undefined

    return (
        <BankersForm submitBonus={this.handleSubmitEntry} joinedName={joinedName} 
                     winnerChosen={winnerChosen} resetGame={this.resetGame} connectionProblem={this.state.connectionProblem}/>
    )
  }
}

BankersFormContainer.propTypes = {
  updateConnectionStatus: PropTypes.func.isRequired,
  changeToDiners: PropTypes.func.isRequired,
  spdzProxyServerList: PropTypes.array.isRequired,
  spdzApiRoot: PropTypes.string.isRequired,
  clientPublicKey: PropTypes.string.isRequired
}

export default BankersFormContainer
