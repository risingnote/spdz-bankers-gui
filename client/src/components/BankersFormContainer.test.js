import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { SocketIO, Server } from 'mock-socket'
import BankersFormContainer from './BankersFormContainer'
import BankersForm from './BankersForm'
import { noProxiesSpdzConfig, twoProxiesWith1Connected, twoProxiesWith2Connected } from '../test_support/ProxyServerList'

const convertToArray = immutList => immutList.toArray().map( immutableMap => immutableMap.toObject())

// Mock out REST call
jest.mock('spdz-gui-lib/dist/rest_api')
import { sendInputsWithShares } from 'spdz-gui-lib/dist/rest_api'

const mockFn = jest.fn()

describe('Bankers Container rendering and behaviour', () => {
  it('Renders as expected (compared to a snapshot) at start of game', () => {
    const tree = renderer.create(
      <BankersFormContainer updateConnectionStatus={mockFn} changeToDiners={mockFn}
                        spdzProxyServerList={convertToArray(twoProxiesWith1Connected)}
                        spdzApiRoot={'/spdzapi'} clientPublicKey={'a1b2c3d4'} />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Renders as expected (compared to a snapshot) at start of game, no proxies found', () => {
    const tree = renderer.create(
      <BankersFormContainer updateConnectionStatus={mockFn} changeToDiners={mockFn}
                        spdzProxyServerList={convertToArray(noProxiesSpdzConfig)}
                        spdzApiRoot={'/spdzapi'} clientPublicKey={'a1b2c3d4'} />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  // Until work out how to include websockets in these tests, just exclude
  xit('Handles the submission of the bonus', () => {
    //Mock out the sockets server
    const mockServer = new Server('http://localhost:3001/diners')
    mockServer.on('joinMeal', (msg, resultCallback) => {
        resultCallback()
    })
    window.io = SocketIO

    sendInputsWithShares.mockImplementation(() => Promise.resolve())

    const wrapper = mount(
      <BankersFormContainer allProxiesConnected={true} spdzProxyServerList={twoProxiesWith2Connected}
                        spdzApiRoot={'/spdzapi'} clientPublicKey={'a1b2c3d4'} />
    )

    const form = wrapper.find(BankersForm)
    const input = form.find('[id="bonusValue"]') //property selector find
    //Simulate user input followed by click submit
    const userInput = 123
    input.simulate('change', {target: {value: userInput}})
    form.simulate('submit')

    expect(sendInputsWithShares.mock.calls.length).toEqual(1)
    expect(sendInputsWithShares.mock.calls[0]).toEqual([[userInput], true, twoProxiesWith2Connected, '/spdzapi', 'a1b2c3d4'])

    sendInputsWithShares.mockClear()
  })
})
