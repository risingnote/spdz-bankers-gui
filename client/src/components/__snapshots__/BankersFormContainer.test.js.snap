exports[`Bankers Container rendering and behaviour Renders as expected (compared to a snapshot) at start of game 1`] = `
<form
  className="BankersForm"
  onSubmit={[Function]}>
  <h4>
    Bankers Celebration Dinner
  </h4>
  <p>
    ...but who should pay?
  </p>
  <label
    htmlFor="joinName">
    Join meal as
  </label>
  <input
    id="joinName"
    onChange={[Function]}
    type="text"
    value="" />
  <label
    htmlFor="bonusValue">
    Bonus
  </label>
  <input
    id="bonusValue"
    onChange={[Function]}
    type="text"
    value="" />
  <label
    className="BankersForm-inline"
    htmlFor="finished">
    Have all joined?
  </label>
  <input
    checked={false}
    className="BankersForm-inline"
    id="finished"
    onChange={[Function]}
    type="checkbox" />
  <p />
  <button
    disabled="disabled"
    style={
      Object {
        "marginRight": "2rem",
      }
    }
    type="submit">
    Send
  </button>
  <button
    disabled="disabled"
    onClick={[Function]}
    type="button">
    Reset
  </button>
  <p
    className="infoMessage">
    
  </p>
</form>
`;

exports[`Bankers Container rendering and behaviour Renders as expected (compared to a snapshot) at start of game, no proxies found 1`] = `
<form
  className="BankersForm"
  onSubmit={[Function]}>
  <h4>
    Bankers Celebration Dinner
  </h4>
  <p>
    ...but who should pay?
  </p>
  <label
    htmlFor="joinName">
    Join meal as
  </label>
  <input
    id="joinName"
    onChange={[Function]}
    type="text"
    value="" />
  <label
    htmlFor="bonusValue">
    Bonus
  </label>
  <input
    id="bonusValue"
    onChange={[Function]}
    type="text"
    value="" />
  <label
    className="BankersForm-inline"
    htmlFor="finished">
    Have all joined?
  </label>
  <input
    checked={false}
    className="BankersForm-inline"
    id="finished"
    onChange={[Function]}
    type="checkbox" />
  <p />
  <button
    disabled="disabled"
    style={
      Object {
        "marginRight": "2rem",
      }
    }
    type="submit">
    Send
  </button>
  <button
    disabled="disabled"
    onClick={[Function]}
    type="button">
    Reset
  </button>
  <p
    className="infoMessage">
    
  </p>
</form>
`;
